### ISSUE TITLE
Auto deny ACC tool account requests if the wiki user doesn't exist

### ISSUE THREAD

**Issue by AfroThundr3007730** [enwikipedia-acc/waca#512]
I imagine we'd get less invalid requests on the internal interface if the tool incorporated a check to Special:CentralAuth to verify that a wiki user actually exists somewhere with that name. Legitimate prospective users that fail the check due to a typo could just try again.

This is a minor problem, to be sure, but still.

**Comment by AfroThundr3007730** [comment]
The tool could also check if the user had > 500 edits and show a warning before submission...

**Comment by stwalkerster** [comment]
This seems like it's essentially a duplicate of #508 ?

**Comment by AfroThundr3007730** [comment]
Yeah, #508 would entirely eliminate the need for this fix. Unless #508 is the more difficult of the two, then this might still be a good fit in the interim.

**Comment by stwalkerster** [comment]
Right; but when would you do the check on the user's information? Currently, we do all the tool account creation stuff, then do the oauth handoff. If the user aborts at the Oauth part, the tool account is still present and still pending approval just with no link to an actual account.

If you want to do it in the signup flow, it will have to be *after* the OAuth has returned, which might never happen. Otherwise, it will need to be a background maintenance job to auto-decline at a later point.

Either way; we're probably needing to redesign the signup workflow with both this and #508.

**Comment by AfroThundr3007730** [comment]
I imagine the check could be a fairly lightweight method, using a [CentralAuth API](https://www.mediawiki.org/wiki/Extension:CentralAuth/API) query done via PHP or JS before form submit.

If I were requesting an account with my alt (`AfroThundr54230`) the app would send a query like so:

    https://meta.wikimedia.org/w/api.php?action=query&meta=globaluserinfo&guiuser=AfroThundr54230&format=json

Which would return this object:

```json
{
  "batchcomplete": "",
  "query": {
    "globaluserinfo": {
      "home": "enwiki",
      "id": 46489736,
      "registration": "2016-05-01T22:22:52Z",
      "name": "AfroThundr54230"
    }
  }
}
```

Whereas a nonexistent account (`AfroThundr1`) would return:

```json
{
  "batchcomplete": "",
  "query": {
    "globaluserinfo": {
      "missing": ""
    }
  }
}
```

If `query.globaluserinfo.missing` exists, the account could be automatically rejected. Or perhaps a better approach would be to check for `query.globalusername.id` or even checking that `query.globalusername.name` matches the requested account (which it should, since we queried it by name...) then allow the submission.

-----

Another check to automate would be the edit count. Consider a user object with all properties (using my alt, which is a much smaller object):

    https://meta.mediawiki.org/w/api.php?action=query&meta=globaluserinfo&guiuser=AfroThundr54230&guiprop=groups|rights|merged|unattached|editcount&format=json

Which returns the object:

```json
{
  "batchcomplete": "",
  "query": {
    "globaluserinfo": {
      "home": "enwiki",
      "id": 46489736,
      "registration": "2016-05-01T22:22:52Z",
      "name": "AfroThundr54230",
      "groups": [],
      "rights": [],
      "merged": [
        {
          "wiki": "enwiki",
          "url": "https://en.wikipedia.org",
          "timestamp": "2016-05-01T22:22:52Z",
          "method": "new",
          "editcount": 10,
          "registration": "2016-05-01T22:22:51Z"
        },
        {
          "wiki": "loginwiki",
          "url": "https://login.wikimedia.org",
          "timestamp": "2016-05-01T22:22:55Z",
          "method": "login",
          "editcount": 0,
          "registration": "2016-05-01T22:22:55Z"
        },
        {
          "wiki": "mediawikiwiki",
          "url": "https://www.mediawiki.org",
          "timestamp": "2016-05-01T22:22:55Z",
          "method": "login",
          "editcount": 0,
          "registration": "2016-05-01T22:22:55Z"
        },
        {
          "wiki": "metawiki",
          "url": "https://meta.wikimedia.org",
          "timestamp": "2016-05-01T22:22:55Z",
          "method": "login",
          "editcount": 4,
          "registration": "2016-05-01T22:22:55Z"
        }
      ],
      "editcount": 14,
      "unattached": []
    }
  }
}
```

Where `query.globaluserinfo.editcount` holds my global edit count (including deleted edits), and `query.globaluserinfo.merged[0].editcount` has my enwiki edit count (in this case at least; the wiki list may appear in a different order).

So the tool could check that `editcount > 500` (either globally or on enwiki) and auto reject the request if that check fails. I envision this could all be done with PHP or JS magic before allowing the submission to complete. What do you think?

**Comment by stwalkerster** [comment]
Yeah, how to check against the API isn't the issue here, it's where to *put* that check. We do not require usernames on the tool match on-wiki usernames, and there are a number of users for whom this is the case.

We don't get an on-wiki username to check for existence until the OAuth flow has completed. If we get an access token back with which we can pull the identity ticket from the API (containing the user's onwiki username), we're already guaranteed that the user account exists onwiki. 

For clarity, this is what happens:
1. User fills in their tool username, a tool password, and a contact email address
2. The user submits the form
3. A tool account is created, marked as "new" and thus appears in the pending approval list, and IRC notifications are sent out.
4. The user is redirected to follow the OAuth flow to link their account
5. The user completes the OAuth flow on-wiki, and is redirected back to the tool with the tool acquiring an access token
6. The tool then requests the user's identity ticket from the wiki using the access token

If the user aborts the OAuth process at step 5, we're left without any idea of the user's username on-wiki to do the check with. At this point, the request is already in the approval queue with a red mark saying that OAuth is not linked.

We can inject code to do the check at step 3, but this is before we have a username with which to check. Alternatively, we can do a check for the user's editcount etc at step 6, but that's after the account has already entered the queue for approval, and doesn't catch 90% of the declined accounts which were requested without linking OAuth.

Alternatively, we can add a sweeper process that runs on a schedule to auto-decline some requests, but that's quite hacky and would invite a race condition for those signing up at the moment the sweeper runs.

Alternatively, we flip the entire signup process around so we can do the checks before the account is created, which is what #508 is trying to do.

We don't even need to do any extra API calls for this - the identity ticket (from `https://en.wikipedia.org/w/index.php?title=Special:OAuth/identify&format=json`) contains all the information we need, including edit count, registration date, whether you're blocked, what groups you are in, and what grants you've given the app. You can see yours from the tool preferences page if you're curious.

**Comment by AfroThundr3007730** [comment]
Ok, then waiting on #508 is probably the best long term solution to this, since it will avoid duplicating the necessary checks before and after OAuth.

FWIW, in the case of users not having the same name on the wiki, you could display a prompt to enter their wiki username and a diff (to prevent impersonation, the contents of which would be a token). That's kind of a hacky method though.

But as you've said, flipping the process would be the ideal solution.

**Comment by stwalkerster** [comment]
Entering a username and a diff ID was actually the process we used before OAuth, and that logic actually still exists in the code behind a config flag. OAuth makes it a **lot** easier though.

**Comment by stwalkerster** [comment]
I think I'm going to close this in favour of #508
